File: backend/app/api/routes/modifiers.py
"""Ürün opsiyonu (modifier) uçları: grup/opsiyon CRUD + ürüne atama."""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status

from app.api.deps import CurrentRestaurant, DbSession, require_manager, require_staff
from app.models.user import User
from app.schemas.modifier import (
    AssignGroupsRequest,
    ModifierCreate,
    ModifierGroupCreate,
    ModifierGroupOut,
    ModifierGroupUpdate,
    ModifierUpdate,
)
from app.services import modifier_service
from app.services.modifier_service import ModifierError

router = APIRouter(tags=["modifiers"])

RequireStaff = Annotated[User, Depends(require_staff)]
RequireManager = Annotated[User, Depends(require_manager)]


def _not_found(exc: ModifierError) -> HTTPException:
    return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))


@router.get("/modifier-groups", response_model=list[ModifierGroupOut])
async def list_groups(
    db: DbSession, restaurant: CurrentRestaurant, _: RequireStaff
) -> list[ModifierGroupOut]:
    groups = await modifier_service.list_groups(db, restaurant.id)
    return [ModifierGroupOut.model_validate(g) for g in groups]


@router.post(
    "/modifier-groups", response_model=ModifierGroupOut, status_code=status.HTTP_201_CREATED
)
async def create_group(
    data: ModifierGroupCreate, db: DbSession, restaurant: CurrentRestaurant, _: RequireManager
) -> ModifierGroupOut:
    group = await modifier_service.create_group(db, restaurant.id, data)
    return ModifierGroupOut.model_validate(group)


@router.patch("/modifier-groups/{group_id}", response_model=ModifierGroupOut)
async def update_group(
    group_id: uuid.UUID,
    data: ModifierGroupUpdate,
    db: DbSession,
    restaurant: CurrentRestaurant,
    _: RequireManager,
) -> ModifierGroupOut:
    try:
        group = await modifier_service.update_group(db, restaurant.id, group_id, data)
    except ModifierError as exc:
        raise _not_found(exc) from exc
    return ModifierGroupOut.model_validate(group)


@router.delete("/modifier-groups/{group_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_group(
    group_id: uuid.UUID, db: DbSession, restaurant: CurrentRestaurant, _: RequireManager
) -> None:
    try:
        await modifier_service.delete_group(db, restaurant.id, group_id)
    except ModifierError as exc:
        raise _not_found(exc) from exc


@router.post("/modifier-groups/{group_id}/modifiers", response_model=ModifierGroupOut)
async def add_modifier(
    group_id: uuid.UUID,
    data: ModifierCreate,
    db: DbSession,
    restaurant: CurrentRestaurant,
    _: RequireManager,
) -> ModifierGroupOut:
    try:
        group = await modifier_service.add_modifier(db, restaurant.id, group_id, data)
    except ModifierError as exc:
        raise _not_found(exc) from exc
    return ModifierGroupOut.model_validate(group)


@router.patch("/modifiers/{modifier_id}", response_model=ModifierGroupOut)
async def update_modifier(
    modifier_id: uuid.UUID,
    data: ModifierUpdate,
    db: DbSession,
    restaurant: CurrentRestaurant,
    _: RequireManager,
) -> ModifierGroupOut:
    try:
        group = await modifier_service.update_modifier(db, restaurant.id, modifier_id, data)
    except ModifierError as exc:
        raise _not_found(exc) from exc
    return ModifierGroupOut.model_validate(group)


@router.delete("/modifiers/{modifier_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_modifier(
    modifier_id: uuid.UUID, db: DbSession, restaurant: CurrentRestaurant, _: RequireManager
) -> None:
    try:
        await modifier_service.delete_modifier(db, restaurant.id, modifier_id)
    except ModifierError as exc:
        raise _not_found(exc) from exc


@router.get("/menu-items/{item_id}/modifier-groups", response_model=list[ModifierGroupOut])
async def get_item_groups(
    item_id: uuid.UUID, db: DbSession, restaurant: CurrentRestaurant, _: RequireStaff
) -> list[ModifierGroupOut]:
    groups = await modifier_service.get_item_groups(db, restaurant.id, item_id)
    return [ModifierGroupOut.model_validate(g) for g in groups]


@router.put("/menu-items/{item_id}/modifier-groups", response_model=list[ModifierGroupOut])
async def assign_item_groups(
    item_id: uuid.UUID,
    data: AssignGroupsRequest,
    db: DbSession,
    restaurant: CurrentRestaurant,
    _: RequireManager,
) -> list[ModifierGroupOut]:
    try:
        groups = await modifier_service.assign_groups(
            db, restaurant.id, item_id, data.group_ids
        )
    except ModifierError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    return [ModifierGroupOut.model_validate(g) for g in groups]

File: frontend/src/shared/api/modifiers.ts
import { apiFetch } from "@shared/api/client";

export interface Modifier {
  id: string;
  modifier_group_id: string;
  name: string;
  price_delta: string;
  is_available: boolean;
  sort_order: number;
}

export interface ModifierGroup {
  id: string;
  restaurant_id: string;
  name: string;
  selection_type: "single" | "multiple";
  min_select: number;
  max_select: number;
  is_required: boolean;
  sort_order: number;
  is_active: boolean;
  modifiers: Modifier[];
}

export interface NewModifier {
  name: string;
  price_delta: number;
}

export interface NewGroup {
  name: string;
  selection_type: "single" | "multiple";
  min_select: number;
  max_select: number;
  is_required: boolean;
  modifiers: NewModifier[];
}

export const modifiersApi = {
  listGroups: () => apiFetch<ModifierGroup[]>("/modifier-groups"),

  createGroup: (data: NewGroup) =>
    apiFetch<ModifierGroup>("/modifier-groups", { method: "POST", body: data }),

  deleteGroup: (id: string) =>
    apiFetch<void>(`/modifier-groups/${id}`, { method: "DELETE" }),

  updateGroup: (
    id: string,
    data: Partial<{
      name: string;
      selection_type: "single" | "multiple";
      is_required: boolean;
      min_select: number;
      max_select: number;
    }>,
  ) => apiFetch<ModifierGroup>(`/modifier-groups/${id}`, { method: "PATCH", body: data }),

  addModifier: (groupId: string, data: NewModifier) =>
    apiFetch<ModifierGroup>(`/modifier-groups/${groupId}/modifiers`, {
      method: "POST",
      body: data,
    }),

  updateModifier: (id: string, data: { name?: string; price_delta?: number }) =>
    apiFetch<ModifierGroup>(`/modifiers/${id}`, { method: "PATCH", body: data }),

  deleteModifier: (id: string) =>
    apiFetch<void>(`/modifiers/${id}`, { method: "DELETE" }),

  itemGroups: (itemId: string) =>
    apiFetch<ModifierGroup[]>(`/menu-items/${itemId}/modifier-groups`),

  assignGroups: (itemId: string, group_ids: string[]) =>
    apiFetch<ModifierGroup[]>(`/menu-items/${itemId}/modifier-groups`, {
      method: "PUT",
      body: { group_ids },
    }),
};
